tsfile/src/main/java/org/apache/iotdb/tsfile/write/writer/TsFileIOWriter.java

```diff
@@ -254,64 +260,51 @@ public void endFile(Schema schema) throws IOException {
-        .entrySet()) {
-      // update statistics in TsDeviceMetadata
-      currentTsDeviceMetadata = entry.getValue();
-
-      // flush tsChunkGroupBlockMetaData
-      offset = out.getPosition();
-      int size = currentTsDeviceMetadata.serializeTo(out.wrapAsStream());
-
-      TsDeviceMetadataIndex tsDeviceMetadataIndex = new TsDeviceMetadataIndex(offset, size,
-          currentTsDeviceMetadata);
-      tsDeviceMetadataIndexMap.put(entry.getKey(), tsDeviceMetadataIndex);
+  private long[] flushAllChunkMetadataList() throws IOException {
+    if (timeseriesMetadataMap.size() == 0) {
+      return new long[0];
     }
-
-    return tsDeviceMetadataIndexMap;
-  }
-
-  /**
-   * group all chunk group metadata by device.
-   *
-   * @param chunkGroupMetaDataList all chunk group metadata
-   * @return TsDeviceMetadata of all devices
-   */
-  private TreeMap<String, TsDeviceMetadata> getAllTsDeviceMetadata(
-      List<ChunkGroupMetaData> chunkGroupMetaDataList) {
-    String currentDevice;
-    TreeMap<String, TsDeviceMetadata> tsDeviceMetadataMap = new TreeMap<>();
-
-    for (ChunkGroupMetaData chunkGroupMetaData : chunkGroupMetaDataList) {
-      currentDevice = chunkGroupMetaData.getDeviceID();
-
-      if (!tsDeviceMetadataMap.containsKey(currentDevice)) {
-        TsDeviceMetadata tsDeviceMetadata = new TsDeviceMetadata();
-        tsDeviceMetadataMap.put(currentDevice, tsDeviceMetadata);
+    // convert ChunkMetadataList to this field
+    long[] tsOffsets = new long[timeseriesMetadataMap.size() + 1];
+    List<TimeseriesMetaData> tsMetadataList = new ArrayList<>();
+    // flush timeseriesMetadataList one by one
+    int i = 0;
+    for (Map.Entry<Path, List<ChunkMetaData>> entry : timeseriesMetadataMap.entrySet()) {
+      Path path = entry.getKey();
+      String deviceId = path.getDevice();
+      if (!deviceOffsetsMap.containsKey(deviceId)) {
+        deviceOffsetsMap.put(deviceId, new int[2]);
+        deviceOffsetsMap.get(deviceId)[0] = i;
+      }
+      deviceOffsetsMap.get(deviceId)[1] = i + 1;
+      TimeseriesMetaData tsMetaData = new TimeseriesMetaData();
+      
+      tsMetaData.setMeasurementId(path.getMeasurement());
+      tsMetaData.setOffsetOfChunkMetaDataList(out.getPosition());
+      int chunkMetadataSize = 0;
+      for (ChunkMetaData chunkMetadata : entry.getValue()) {
+        chunkMetadataSize += chunkMetadata.serializeTo(out.wrapAsStream());
       }
-      tsDeviceMetadataMap.get(currentDevice).addChunkGroupMetaData(chunkGroupMetaData);
+      tsMetaData.setDataSizeOfChunkMetaDataList(chunkMetadataSize);
+      tsMetaData.setNumOfChunkMetaDatas(entry.getValue().size());
+      tsMetadataList.add(tsMetaData);
+      i++;
     }
-    return tsDeviceMetadataMap;
+    
+    for (i = 0; i < tsMetadataList.size(); i++) {
+      tsOffsets[i] = out.getPosition();
+      int size = tsMetadataList.get(i).serializeTo(out.wrapAsStream());
+      tsOffsets[i+1] = tsOffsets[i] + size;
+    }
```

Comment:
Please remove useless assignment and addition operations.
```
if(i > 0){
      tsOffsets[0] = out.getPosition();
}
for (i = 0; i < tsMetadataList.size(); i++) {
      int size = tsMetadataList.get(i).serializeTo(out.wrapAsStream());
      tsOffsets[i+1] = out.getPosition();
 }
```

Reply:
The tsOffsets array has been removed.